src/utils/gauss_quad.hpp

```diff
@@ -0,0 +1,657 @@
+// FITNESS FOR A PARTICULAR PURPOSE, TITLE AND NON-INFRINGEMENT. IN NO EVENT
+// SHALL THE COPYRIGHT HOLDERS OR ANYONE DISTRIBUTING THE SOFTWARE BE LIABLE
+// FOR ANY DAMAGES OR OTHER LIABILITY, WHETHER IN CONTRACT, TORT OR OTHERWISE,
+// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//========================================================================================
+// Parthenon performance portable AMR framework
+// Copyright(C) 2020 The Parthenon collaboration
+// Licensed under the 3-clause BSD License, see LICENSE file for details
+//========================================================================================
+// (C) (or copyright) 2021. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#ifndef PARTHENON_UTILS_GAUSS_QUAD_HPP_
+#define PARTHENON_UTILS_GAUSS_QUAD_HPP_
+
+#include <array>
+#include <cmath>
+#include <limits>
+
+#include "../kokkos_abstraction.hpp"
+
+namespace parthenon::math::quadrature {
+namespace detail {
+
+enum class precision { Invalid, Float, Double, LongDouble };
+
+template <class T>
+struct gauss_constant_category {
+  static constexpr auto value =
+      (std::numeric_limits<T>::is_specialized == 0) ? precision::Invalid
+      : (std::numeric_limits<T>::radix == 2)
+          ? ((std::numeric_limits<T>::digits <= std::numeric_limits<float>::digits) &&
+                     std::is_convertible<float, T>::value
+                 ? precision::Float
+             : (std::numeric_limits<T>::digits <= std::numeric_limits<double>::digits) &&
+                     std::is_convertible<double, T>::value
+                 ? precision::Double
+             : (std::numeric_limits<T>::digits <=
+                std::numeric_limits<long double>::digits) &&
+                     std::is_convertible<long double, T>::value
+                 ? precision::LongDouble
+                 : precision::Invalid)
+          : precision::Invalid;
+};
+
+template <class Real, unsigned N, precision>
+class gauss_detail;
+
+template <class T>
+class gauss_detail<T, 7, precision::Float> {
+ public:
+  KOKKOS_FUNCTION static auto abscissa() -> std::array<T, 4> const & {
```

Comment:
All of the `abcissa()` and `weights()` functions need to be marked `KOKKOS_FORCEINLINE_FUNCTION`, otherwise the stack will overflow and the GPU kernel will crash :)